Incremental Search: extract logic for disabling on RegEx-Search #1791

Wittmaxi opened the pull request:
when "incremental search" is selected and "RegEx-Search" is as well, the dialog/overlay will disable the option for "whole words". This PR extracts that functionality into the Find/Replace-Logic.

Extracted from a Review-Comment in #1791

Commit:
Move activation logic for SearchOptions.INCREMENTAL to FindReplaceLogic

when "incremental search" is selected and "RegEx-Search" is as well,
the dialog/overlay will disable the option for "whole
words". This PR extracts that functionality into the Find/Replace-Logic.

Extracted from a Review-Comment in #1791

Wittmaxi commented:
Random Fail for Ubuntu is reported here: https://github.com/eclipse-platform/eclipse.platform.ui/issues/1808
Random Fail for Mac is reported here: https://github.com/eclipse-platform/eclipse.platform.ui/issues/811
Random Fail for Windows is reported here: https://github.com/eclipse-platform/eclipse.platform.ui/issues/1427

HeikoKlare commented on tests/org.eclipse.ui.workbench.texteditor.tests/src/org/eclipse/ui/workbench/texteditor/tests/FindReplaceDialogTest.java:
```diff
@@ -392,6 +393,19 @@ public void testReplaceAndFindAfterInitializingFindWithSelectedString() {
+
+		dialog.incrementalCheckBox.setSelection(true);
+		dialog.regExCheckBox.setSelection(true);
+		dialog.regExCheckBox.notifyListeners(SWT.Selection, null);
```
Please use the existing `select()` method instead, which already does the combination of setting the selection and informing listeners.

Wittmaxi replied:
Done in https://github.com/eclipse-platform/eclipse.platform.ui/commit/d654b06728ceb6865a378e7608c0fec4e1ed13f3

HeikoKlare commented on bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/IFindReplaceLogic.java:
```diff
@@ -66,6 +66,14 @@ public interface IFindReplaceLogic {
+	 * Returns whether the Find/Replace-Overlay may perform incremental search
+	 * currently.
+	 *
+	 * @return whether incremental search may be performed with the current options.
```
This is the logic interface, which is completely independent from UI / overlay. If a method is only a getter (and thus only needs a "returns" documentation), you can also use a simplified Javadoc syntax: `{@return ... }`
```suggestion
	 * @return whether incremental search may be performed based on the currently active options
```

Wittmaxi replied:
I am not quite sure what you mean with that syntax?
I did adapt your JavaDoc-text change

HeikoKlare replied:
The javadoc still refers to UI. But unfortunately my proposal was incomplete, sorry for that. This is what I would propose to have the javadoc look like:

```java
/**
  * {@return whether incremental search may be performed based on the currently active options}
  */
```

Then it renders as follows:
<img width="544" alt="image" src="https://github.com/eclipse-platform/eclipse.platform.ui/assets/755472/f0c25435-84fb-471a-8f5b-6ce98b096976">

Wittmaxi replied:
I didn't know this was possible. Thank you for the suggestion!

HeikoKlare reviewed (commented):
In general, this looks good. I have two minor comments.

HeikoKlare commented:
And could you please also adapt the commit message, just like for the "whole words" options? (see https://github.com/eclipse-platform/eclipse.platform.ui/pull/1791#issuecomment-2046929929)

Wittmaxi commented:
@HeikoKlare Addressed in https://github.com/eclipse-platform/eclipse.platform.ui/commit/d654b06728ceb6865a378e7608c0fec4e1ed13f3

fedejeanne commented on tests/org.eclipse.ui.workbench.texteditor.tests/src/org/eclipse/ui/workbench/texteditor/tests/FindReplaceDialogTest.java:
```diff
@@ -414,6 +415,18 @@ public void testActivateWholeWordsAndSearchForTwoWords() {
 	}
 
+	@Test
+	public void testIncrementalSearchOnlyActiveWhenAllowed() {
```
The name of the method mentions "`active`" but it actually checks whether or not the checkbox is **enabled**. Either the name is wrong or the assertions.

Since I don't see any method in `FindReplaceDialog` that can give you the state of `findReplaceLogic.isIncrementalSearchAvailable()` for this test to check it, I assume a better name for the test would be:
```suggestion
	public void testIncrementalSearchOnlyEnabledWhenAllowed() {
```

Wittmaxi replied:
https://github.com/eclipse-platform/eclipse.platform.ui/commit/859fef9745d8b5af0d1c8d0a5874a5236329b0cd

Good catch! Although now this becomes inconsistent with `testActivateWholeWordsAndSearchForTwoWords` - I wonder if it makes sense to open a new PR to also change that one (or is it fine changing it in this PR?)

fedejeanne replied:
I'd split that one into 2 (in another PR):
* `testWholeWordsOnlyEnabledWhenSearchingForOneWord`
* `testWholeWordsDisnabledWhenSearchingForRegEx`

The test does more than it says anyway, which is confusing

fedejeanne reviewed (changes requested):
Minor improvement in a test name desired

HeikoKlare reviewed (changes requested):
The current state produces a regression, as the dialog can loose its "incremental" option when reopened. To reproduce, activate the "incremental" option, activate and deactive the "regex" option, then close and re-open the dialog. The "incremental" option is unselected but it selection should be recovered from the previous session of the dialog.

This is because the change misses an update of the functionality in  `FindReplaceDialog.setupFindReplaceLogic()`, which still bases the activation of the "incremental" option on its selection _and_ its enablement. The latter has to be removed, just like for the "whole word" option.
https://github.com/eclipse-platform/eclipse.platform.ui/blob/781d89e41f9de47cf7e0d81a0184d34782d69d00/bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/FindReplaceDialog.java#L1267-L1268

fedejeanne commented:
> The current state produces a regression ...

Now that you mention it, I'm pretty sure one can find some other weird behaviors when playing with "Incremental", "Whole words" and "Regular expressions" and closing + reopening the search dialog :-o

HeikoKlare commented:
> Now that you mention it, I'm pretty sure one can find some other weird behaviors when playing with "Incremental", "Whole words" and "Regular expressions" and closing + reopening the search dialog :-o

Can you elaborate on that? Is there something specific you have in mind and is that a general negative expectation or related to this PR?

fedejeanne commented:
> > Now that you mention it, I'm pretty sure one can find some other weird behaviors when playing with "Incremental", "Whole words" and "Regular expressions" and closing + reopening the search dialog :-o
> 
> Can you elaborate on that? Is there something specific you have in mind and is that a general negative expectation or related to this PR?

Sorry for being unclear. I wasn't talking about this PR :-)

Wittmaxi commented:
@HeikoKlare I have addressed your comment and added a Unit-Test which checks the behavior for "re-opening" the dialog.
A very large amount of improvements in testing-architecture were made in the Find/Replace-Overlay-PR, so once that is moved, we should test the "re-opening"-behavior of the Dialog/Overlay with the other buttons

Wittmaxi commented:
Windows Random Failing documented here: https://github.com/eclipse-platform/eclipse.platform.ui/issues/1807

HeikoKlare commented on tests/org.eclipse.ui.workbench.texteditor.tests/src/org/eclipse/ui/workbench/texteditor/tests/FindReplaceDialogTest.java:
```diff
@@ -148,12 +151,16 @@ private void assertInitialConfiguration() {
+		void closeAndRestore() {
 			restoreInitialConfiguration();
 			assertInitialConfiguration();
 			closeOperation.run();
```
Just a nitpicky final comment. 
```suggestion
			close();
```

HeikoKlare replied:
Done in a767ca3

HeikoKlare commented:
Failing checks because of new issues are only because of an outdated reference build. Recent master build reports the same new issues: https://ci.eclipse.org/platform/job/eclipse.platform.ui/job/master/422
